api4 explorer: make boolean params work in cv (short syntax)

highfalutin opened the pull request:
Overview
----------------------------------------
Fixes https://lab.civicrm.org/dev/core/-/issues/4129

Before
----------------------------------------
In API v4 explorer, boolean params are rendered in CV (short) syntax as `true` or `false`. The generated code does not work in `cv`.

After
----------------------------------------
Params are rendered as `1` or `0`. Generated code works in `cv`.

highfalutin committed:
api4 explorer: make boolean params work in cv (short syntax)

highfalutin commented:
Well actually in some cases the `cv` command still doesn't work... `cv` interprets all param values given in key=value syntax as strings. Even if we pass, for example, `useTrash=0` instead of `useTrash=false`, the param value will still be **string** `"0"` and so [type-checking will fail](https://github.com/civicrm/civicrm-core/blob/af1fdfa1af4749fb25dbe225d0c561f986e2251f/Civi/Api4/Event/Subscriber/ValidateFieldsSubscriber.php#L51) and throw an error.

In at least one case ([\Civi\Api4\Generic\AbstractAction::setCheckPermissions()](https://github.com/civicrm/civicrm-core/blob/6ef03256261c60a017a4feb23d95446771802037/Civi/Api4/Generic/AbstractAction.php#L192)), the string is explicitly recast to a boolean, and this works correctly for `checkPermissions=0`. But for `checkPermissions=false`, **the string `"false"` is recast to boolean `true`**, and no error is thrown (except maybe "Authorization Failed").

totten commented:
@highfalutin I think your original intuition is pretty good -- passing `0`/`1` on CLI generally works better than passing `"false"`/`"true"`. This is historically true in similar interfaces (like `cv api3` and `drush cvapi`); it's true in related ecosystems (like C and MySQL); and it checks out for several things in `cv api4`:

* `checkPermission=0`: Numeric-bool works. String-bool doesn't.
* `+v do_not_sms=1`: Numeric-bool works correctly. String-bool doesn't.
* `+w do_not_phone=0`: Numeric-bool works correctly. String-bool doesn't.

That is curious that `useTrash` differs. I think the difference between `useTrash` and `checkPermissions` is this:

* `$checkPermissions` has a [real setter method with a PHP type declaration](https://github.com/civicrm/civicrm-core/blob/6ef03256261c60a017a4feb23d95446771802037/Civi/Api4/Generic/AbstractAction.php#L192).
    * When one calls `setCheckPermissions(1)`, it applies PHP type-casting rules (ie `1` ==> `true`).
    * It stores `$this->checkPermissions=true`.
    * Later on, [Api4's ValidateFieldsSubscriber](https://github.com/civicrm/civicrm-core/blob/af1fdfa1af4749fb25dbe225d0c561f986e2251f/Civi/Api4/Event/Subscriber/ValidateFieldsSubscriber.php#L50) accepts this.
* `$useTrash`  has a [magic setter method with no PHP type declaration](https://github.com/civicrm/civicrm-core/blob/65c45021eff335e59854a86d2697f1ba7d742675/Civi/Api4/Generic/Traits/SoftDeleteActionTrait.php#L16).
    * When you call `setUseTrash(1)`, it will pass through any `mixed` value without validation or casting.
    * It stores `$this->useTrash` as `1`.
    * Later on, [Api4's ValidateFieldsSubscriber](https://github.com/civicrm/civicrm-core/blob/af1fdfa1af4749fb25dbe225d0c561f986e2251f/Civi/Api4/Event/Subscriber/ValidateFieldsSubscriber.php#L50) rejects this.

(To confirm this is the difference: you can [add a real setter method for `useTrash`](https://gist.github.com/totten/d9a5fafa6a788b13cf0cbb653cbfc47b), and it `cv api4 Foo.delete useTrash=0` will work.)

---

IMHO, this patch is better than what it was doing before. :+1: 

Aside: There is a related issue in how the explorer renders booleans in the "Where" (`+w`) and "Value" (`+v`) clauses. But that doesn't have to stop this.

eileenmcnaughton commented:
I added merge-ready based on @totten's "IMHO, this patch is better than what it was doing before. "

I think some related issues came up - but they don't need to be in scope for this PR so I'll just let @highfalutin have a chance to read & respond

highfalutin commented:
@totten thanks for the comments, and thanks for #25595 which goes a long way to addressing the issues I mentioned in https://github.com/civicrm/civicrm-core/pull/25589#issuecomment-1432331555 above.

In addition to the `+where` problem you mentioned, I think it's also not ideal for strings to get turned into booleans silently in some cases (I'm looking at you, `setCheckPermissions()`), but none of that should prevent this PR from getting merged.
